src/app/qgshandlebadlayers.cpp

```diff
@@ -509,3 +479,161 @@ int QgsHandleBadLayers::layerCount()
 {
   return mLayerList->rowCount();
 }
+
+QString QgsHandleBadLayers::findFile( const QString &fileName, const QString &basePath, int maxDepth, const int driveMargin )
+{
+  int depth = 0;
+  QDir folder = QDir( basePath );
+  QString existingBase;
+  while ( !folder.exists() && folder.absolutePath().count( folder.separator() ) > driveMargin )
+  {
+    existingBase = folder.path();
+    if ( existingBase.contains( '/' ) )
+    {
+      existingBase = existingBase.left( existingBase.lastIndexOf( '/' ) );
+      folder = QDir( existingBase );
+    }
+    else
+      folder.cdUp();
+    depth += 1;
+  }
+  if ( depth > maxDepth )
+    maxDepth = depth;
+  while ( depth <= maxDepth && folder.cdUp() && folder.absolutePath().count( folder.separator() ) > driveMargin )
+  {
+    QDirIterator finder( folder.path(), QStringList() << fileName, QDir::Files, QDirIterator::Subdirectories );
+    if ( finder.hasNext() )
+      return ( finder.next() );
+    else
+      depth += 1;
+  }
+  return ( existingBase + folder.separator() + fileName );
+}
+
+QString QgsHandleBadLayers::checkBasepath( const QString &name, const QString &newPath, const QString &fileName )
+{
+  const QString originalBase = mOriginalFileBase.value( name );
+
+  if ( QFileInfo::exists( newPath ) && QFileInfo( newPath ).isFile() )
+  {
+    const QString newBasepath = newPath.left( newPath.lastIndexOf( '/' ) );
+    if ( !mAlternativeBasepaths.value( originalBase ).contains( newBasepath ) )
+      mAlternativeBasepaths[ originalBase ].append( newBasepath );
+    return ( newPath );
+  }
+  else if ( mAlternativeBasepaths.contains( originalBase ) )
+  {
+    const QList<QString> altPaths = mAlternativeBasepaths.value( originalBase );
+    if ( ! altPaths.isEmpty() )
```

Comment:
No reason for this check I guess?

Reply:
Prevent iteration on an empty list.

Reply:
Iterating an empty list will result in no iteration at all :)

Reply:
Very well, I just wanted to be more safe than segfault.